File: src/bac_generator/schemas/submission.py
from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, ConfigDict


class SubmissionStatus(StrEnum):
    PASSED = "passed"
    PARTIAL = "partial"
    FAILED = "failed"
    COMPILATION_ERROR = "compilation_error"
    RUNTIME_ERROR = "runtime_error"


class SubmissionEvaluation(BaseModel):
    score: int
    passed_tests: int
    total_tests: int
    status: SubmissionStatus
    feedback: str | None = None


class SubmissionProgress(BaseModel):
    has_submitted: bool
    latest_score: int | None
    submission_count: int
    completed: bool
    

class SubmitSolutionRequest(BaseModel):
    code: str


class SubmissionCreate(BaseModel):
    exercise_id: int
    user_id: str
    code: str
    score: int
    passed_tests: int
    total_tests: int
    status: SubmissionStatus
    feedback: str | None = None


class SubmissionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    exercise_id: int
    score: int
    passed_tests: int
    total_tests: int
    status: SubmissionStatus
    feedback: str | None
    created_at: datetime


class OfficialSolutionRead(BaseModel):
    solution: str
    explanation: str

File: frontend/schemas/submission.ts
import { z } from "zod";


export const KNOWN_SUBMISSION_STATUSES = [
  "passed",
  "partial",
  "failed",
  "compilation_error",
  "runtime_error",
] as const;

export type KnownSubmissionStatus = (typeof KNOWN_SUBMISSION_STATUSES)[number];

export const SubmissionSchema = z.object({
  id: z.number(),
  exercise_id: z.number(),
  score: z.number(),
  passed_tests: z.number(),
  total_tests: z.number(),
  status: z.string(),
  feedback: z.string().nullable().optional(),
  created_at: z.string(),
});

export const SubmitSolutionRequestSchema = z.object({
  code: z.string().trim().min(1, "Scrie o soluție înainte de a trimite."),
});


export const OfficialSolutionSchema = z.object({
  solution: z.string(),
  explanation: z.string(),
});

export type Submission = z.infer<typeof SubmissionSchema>;
export type SubmitSolutionRequest = z.infer<typeof SubmitSolutionRequestSchema>;
export type OfficialSolution = z.infer<typeof OfficialSolutionSchema>;
